File: EHaskins.Frc.Communication.Shared/FrcPacketUtils.cs
using System;
using System.Net;
using MiscUtil.Conversion;

public static class FrcPacketUtils
{

    public static bool IsValidFrcPacket(this byte[] data)
    {
        uint dataCrc = EndianBitConverter.Big.ToUInt32(data, data.Length - 4);

        for (int i = data.Length - 4; i < data.Length; i++)
        {
            data[i] = 0;
        }
        //Dim calculatedCrc = BitConverter.ToUInt32(calulatedCrcBytes, 0)
        var calculatedCrc = Crc32.Compute(data);
        return dataCrc == calculatedCrc;
    }

    public static IPAddress GetIP(byte network, ushort teamNumber, byte hostNumber)
    {
        if (teamNumber > 9999)
            throw new ArgumentOutOfRangeException("teamNumber must be less than or equal to 9999.");
        var data = new byte[4];
        data[0] = network;
        data[1] = (byte)(teamNumber / 100);
        data[2] = (byte)(teamNumber % 100);
        data[3] = hostNumber;

        return new IPAddress(data);
    }
}
public enum Devices : byte
{
    Robot = 2,
    PC = 6,
    DS = 5,
    Router = 4,
    GameAdapter = 1
}

File: EHaskins.Frc.Dashboard.Old/DS Shared/Packet Utilities/FrcPacketUtils.vb
Imports System.Net
Imports MiscUtil.Conversion

Public Module FrcPacketUtils
    Public Function VerifyFrcCrc(ByVal data As Byte()) As Boolean
        'Dim calculatedCrc As UInteger = 0
        Dim dataCrc As UInteger = 0
        Dim crc As New Crc32()

        Dim dataCrcBytes As Byte() = data.Skip(data.Length - 4).Take(4).ToArray()
        dataCrc = EndianBitConverter.Big.ToUInt32(dataCrcBytes, 0)

        'remove CRC bytes from data before calculating CRC.
        Dim crcData(data.Length - 1) As Byte
        data.Take(data.Length - 4).ToArray.CopyTo(crcData, 0)

        Dim calulatedCrcBytes As Byte() = crc.ComputeHash(crcData)
        'Dim calculatedCrc = BitConverter.ToUInt32(calulatedCrcBytes, 0)
        Dim calculatedCrc = Crc32.Compute(crcData)
        Return dataCrc = calculatedCrc
    End Function

    Public Function GetIP(ByVal teamNumber As Integer, ByVal device As Devices) As IPAddress
        Dim num2 = teamNumber Mod 100
        Dim num1 = (teamNumber - num2) / 100
        Dim ipStr = String.Format("10.{0}.{1}.{2}", num1, num2, CInt(device))
        Return IPAddress.Parse(ipStr)
    End Function
End Module
Public Enum Devices
    Robot = 2
    PC = 6
    DS = 5
    Router = 4
    GameAdapter = 1
End Enum